Create reviews front end and integrate with back end

MilesDev91 opened the pull request:
### What this PR does:
- Reviews are now retrieved from the database and displayed on profile details page. The first three are displayed in the right grid section, if there are more a button will appear that shows a modal with all reviews.
- Review form was added below the review display section. Rating (default 3) is required but text is optional. It is fully integrated to create a review on the back end.
- Ratings are calculated when retrieving reviews.

### Screenshots / Videos:
![image](https://user-images.githubusercontent.com/34494229/152681529-22f9753a-9338-47d3-961b-4672be88e2da.png)
![image](https://user-images.githubusercontent.com/34494229/152681539-4e6bc9bc-c555-47b5-b170-ee43ae5afea0.png)
![image](https://user-images.githubusercontent.com/34494229/152681577-cb725756-1f45-44e9-a759-6e93374e055c.png)
![image](https://user-images.githubusercontent.com/34494229/152681599-6405cfc1-2551-4adc-9adc-eeff4ead8fc9.png)

### Any information needed to test this feature:
- You must have at least 1 sitter and 4 non sitter accounts to see all parts of this feature. You will need to add ratings to the sitter account if it was created before this PR. Default is 2.5 in the Profile model.
- Navigate to the profile details page of the sitter account (either directly with /profile/{profileId} or by searching on the listings page).
- Create a review on the review form. You should see the review appear on the page. Comments are optional.
- Repeat the previous two steps on at least 3 different accounts to see the modal button appear (you can't create multiple on the same profile with the same account).
- Try to create two reviews on the same account to see validation error.

closes #15

MilesDev91 committed:
Reviews FE

Prepared ProfileDetails front end for reviews. There is a preview section for up to three reviews and, if there are any more, a button that brings up a modal to show all reviews.

Added a Review component, ReviewDialog (modal) component, and Review interface.

Made a small adjustment to review controller to prepare for populating name and photo url.

MilesDev91 committed:
Integration

ProfileDetails now fetches reviews (and their reviewer profiles) from backend.

Started on review form.

MilesDev91 committed:
Review form

Finished review form and prepared it for integration with the backend.

MilesDev91 committed:
Create reviews

Review form on profile details page now creates a review on the backend. Review is added to state and displayed immediately for better user feedback.

Fixed problem with passing profile name and photo to review.

MilesDev91 committed:
Ratings

Added ratings to the profile model. Rating from model is displayed on front end.

Creating a review recalculates the ratings for a profile on the back end.

MilesDev91 committed:
Merge branch 'reviews-be' into reviews-fe-integration

MilesDev91 committed:
Update review.js

MilesDev91 committed:
Validation and Error

MilesDev91 committed:
Update ProfileListing.tsx

Reverted changes.

MilesDev91 committed:
Rating

Calculate rating and return on the back end.

rajivtitus commented on client/src/components/ReviewsDialog/ReviewsDialog.tsx:
```diff
@@ -0,0 +1,30 @@
+      <List>
+        {reviews.map((review, i) => {
+          return (
+            <Box sx={{ padding: `${theme.spacing(2)} ${theme.spacing(2)}` }} key={i}>
```
The usage of indexes as keys is an anti-pattern in React, please use unique keys instead

rajivtitus commented on client/src/components/ProfileReview/ProfileReview.tsx:
```diff
@@ -0,0 +1,25 @@
+  review: Review;
+}
+
+export default function ProfileReview({ review }: Props): JSX.Element {
```
👍🏽

rajivtitus commented on client/src/components/ProfileReview/ProfileReview.tsx:
```diff
@@ -0,0 +1,25 @@
+            photoUrl={review.reviewer.photo}
+          ></AvatarDisplay>
+        </ListItemAvatar>
+        {review.text ? <ListItemText>{review.text}</ListItemText> : null}
```
```suggestion
        {review.text && <ListItemText>{review.text}</ListItemText>}
```
Nit: You can use the short-circuit operator here

MilesDev91 replied:
Oh nice didn't know about that

rajivtitus commented on client/src/components/ProfileReview/ProfileReview.tsx:
```diff
@@ -0,0 +1,25 @@
+        <ListItemAvatar>
+          <AvatarDisplay
+            loggedIn
+            user={{ name: review.reviewer.name, email: 'example@example.com' }}
```
Why are we passing a hardcoded email as a prop here? :)

MilesDev91 replied:
AvatarDisplay is using the User interface which requires an email despite not using the email for anything. Instead of changing either and breaking a bunch of code I thought this would be alright. The root issue should probably be addressed instead though.

rajivtitus commented on client/src/pages/ProfileDetails/ProfileDetails.tsx:
```diff
@@ -29,6 +40,15 @@ export default function ProfileDetails(): JSX.Element {
       getProfile(profileId).then((res) => {
         if (!res.error) {
           setProfile(res.success.profile);
+          getReviews(profileId).then((res) => {
```
Let's also chain a `catch` block to handle a promise rejection.

Alternatively, it would be easier to use async/await syntax for increased readability and maintainability

MilesDev91 replied:
I thought that the API function was handling promise rejection with its own catch block (and setting an error), then we handle whether it returned a success or error in the component. This is how we have been doing it on the other API helper calls, am I missing something in this particular instance?

MilesDev91 replied:
I like using await in this scenario, however, so I might be able to do both API helpers that way and just improve the page overall.

rajivtitus commented on client/src/pages/ProfileDetails/ProfileDetails.tsx:
```diff
@@ -127,8 +164,47 @@ export default function ProfileDetails(): JSX.Element {
+                {reviews && reviews.length > 0 ? (
+                  reviews.slice(0, 3).map((review, i) => {
+                    return (
+                      <Box key={i} className={classes.review}>
```
Same as comment above about usage of indexes as keys

rajivtitus commented on client/src/pages/ProfileDetails/ReviewForm/ReviewForm.tsx:
```diff
@@ -0,0 +1,93 @@
+  };
+
+  return (
+    <Formik
```
💯

rajivtitus commented on client/src/pages/ProfileDetails/ReviewForm/ReviewForm.tsx:
```diff
@@ -0,0 +1,93 @@
+  addReview: (review: Review) => void;
+}
+
+export default function ReviewForm({ profileId, addReview }: Props): JSX.Element {
```
Really good work with this component 🚀

rajivtitus commented on client/src/pages/ProfileDetails/useStyles.ts:
```diff
@@ -27,6 +27,19 @@ const useStyles = makeStyles((theme: Theme) => ({
     marginRight: 20,
   },
+  reviewsContainer: {
+    display: 'flex',
```
These flexbox styles can be added to the Box component directly. 

Personally, for any significant styles I apply them as props to the MUI component, for any other styles I use a CSS class.

rajivtitus commented on client/src/pages/ProfileDetails/ProfileDetails.tsx:
```diff
@@ -48,6 +68,22 @@ export default function ProfileDetails(): JSX.Element {
+
+  const addReview = (review: Review) => {
+    if (reviews) {
+      setReviews([review, ...reviews]);
```
```suggestion
      setReviews((prevReviews) => ([...prevReviews, review]));
```
Whenever we need to update our current state to next state based on value possed by property just now and all this is happening in async fashion it is good idea to use setState as callback function.

rajivtitus reviewed (commented):
Great work with this! Some comments to have a look and address 👍🏽

MilesDev91 committed:
Merge branch 'reviews-be' into reviews-fe-integration

MilesDev91 committed:
Finished adjustments for Reviews BE

MilesDev91 committed:
Minor fixes

Fixed a couple things for React best practices

Removed the useTheme hook from the page, since theme spacing is implied when using a number.

MilesDev91 committed:
Merge branch 'main' into reviews-fe-integration

MilesDev91 committed:
Bookings

Fixed a couple quick things with bookings

MilesDev91 committed:
AvatarDisplay changes

AvatarDisplay doesn't need an email but requires a User interface as a prop (which contains an email). Changed to require a name instead. Fixed places where it was used.

MilesDev91 committed:
Update ProfileDetails.tsx

Switched to await syntax for better readability.
